slatedb/src/record/store.rs

```diff
@@ -0,0 +1,365 @@
+                            continue;
+                        }
+                        Err(err) => return Err(err),
+                        Ok(()) => {
+                            return Ok(Self {
+                                stored,
+                                local_epoch,
+                                get_epoch,
+                                set_epoch,
+                            })
+                        }
+                    }
+                }
+            },
+        )
+        .await
+        .map_err(|_| SlateDBError::Timeout {
+            op: "record update",
+            backoff: Duration::from_secs(1),
+        })
+    }
+
+    #[cfg_attr(not(test), allow(dead_code))]
+    pub(crate) fn local_epoch(&self) -> u64 {
+        self.local_epoch
+    }
+
+    pub(crate) fn id(&self) -> u64 {
+        self.stored.id()
+    }
+
+    pub(crate) fn next_id(&self) -> u64 {
+        self.stored.next_id()
+    }
+
+    pub(crate) fn record(&self) -> &T {
+        self.stored.record()
+    }
+
+    pub(crate) fn clock_arc(&self) -> Arc<dyn SystemClock> {
+        self.stored.clock_arc()
+    }
+
+    pub(crate) fn prepare_dirty(&self) -> Result<DirtyRecord<T>, SlateDBError> {
+        self.check_epoch()?;
+        let id = self.id();
+        let record = self.record().clone();
+        Ok(DirtyRecord::new(id, record))
+    }
+
+    // The file may have been updated by a readers, another process, or
+    // we may have gotten this error after successfully updating
+    // if we failed to get the response. Either way, refresh
+    // the file and try the bump again.
+    pub(crate) async fn refresh(&mut self) -> Result<(), SlateDBError> {
+        self.stored.refresh().await?;
+        self.check_epoch()
+    }
+
+    pub(crate) async fn update_dirty(&mut self, dirty: DirtyRecord<T>) -> Result<(), SlateDBError> {
+        self.check_epoch()?;
+        self.stored.update_dirty(dirty).await
+    }
+
+    #[allow(clippy::panic)]
+    fn check_epoch(&self) -> Result<(), SlateDBError> {
+        let stored_epoch = (self.get_epoch)(self.stored.record());
+        if self.local_epoch < stored_epoch {
+            return Err(SlateDBError::Fenced);
+        }
+        if self.local_epoch > stored_epoch {
+            panic!("the stored epoch is lower than the local epoch");
+        }
+        Ok(())
+    }
+}
+
+impl<T: Clone> StoredRecord<T> {
+    pub(crate) async fn init(store: Arc<RecordStore<T>>, value: T) -> Result<Self, SlateDBError> {
+        store.write(1, &value).await?;
+        Ok(Self {
+            id: 1,
+            record: value,
+            store,
+        })
+    }
+
+    pub(crate) fn new(id: u64, record: T, store: Arc<RecordStore<T>>) -> Self {
+        Self { id, record, store }
+    }
+
+    pub(crate) fn id(&self) -> u64 {
+        self.id
+    }
+    pub(crate) fn record(&self) -> &T {
+        &self.record
+    }
+
+    pub(crate) fn clock_arc(&self) -> Arc<dyn SystemClock> {
```

Comment:
This feels off to me. IMO, it should be removed. FenceableManifest already takes `system_clock`. It should hold on to it and use it rather than pulling the clock back out of its inner record.